### PR TITLE
Model distillation

### PR DESCRIPTION
**Pull request by EricMichaelSmith**
**Patch description**
Open-sources agents for performing knowledge distillation on `transformer/generator` and `bert` models. This requires a small change to `MultiHeadAttention`, to pass back the query/key product matrix used for measuring attention losses in TinyBERT-style distillation.

**Testing steps**
`tests/nightly/gpu/test_anti_scaling.py`

### CODE REVIEW

File: projects/anti_scaling/distillation.py [hunk 1]
```diff
@@ -0,0 +1,986 @@
...
+        num_tokens: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Compute the embedding loss for either the encoder or the decoder.
+        """
+        max_value = torch.finfo(student_emb_output.dtype).max
+        raw_loss = F.mse_loss(
+            input=student_emb_output, target=teacher_emb_output, reduction='none'
+        )
+        clamped_loss = torch.clamp(raw_loss, min=0, max=max_value)
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:502]: Yeah, clamping is more necessary for fp16 to prevent the loss terms from being inf. I'll add a comment about that

File: projects/anti_scaling/distillation.py [hunk 2]
```diff
@@ -158,34 +158,34 @@ def __init__(self, opt, shared=None):
...
         if shared is None:
-            to_copy = {'no_cuda', 'model_parallel', 'fp16', 'fp16_impl'}
+            to_copy = {'no_cuda', 'model_parallel', 'fp16', 'fp16_impl', 'gpu'}
             override = {k: opt[k] for k in to_copy}
             override['datatype'] = 'train:evalmode'  # Don't initialize the optimizer
             teacher_agent = create_agent_from_model_file(opt['teacher_model'], override)
             self.teacher_model = teacher_agent.model
             assert teacher_agent.opt['n_heads'] == opt['n_heads']
-            if hasattr(self.teacher_model, 'module'):
-                self.teacher_model = self.teacher_model.module
```
Comment (stephenroller) [inline projects/anti_scaling/distillation.py]: This new change seems surprising to me?

File: projects/anti_scaling/distillation.py [hunk 3]
```diff
@@ -871,8 +895,13 @@ def compute_loss(self, batch, return_output=False):
 
         fwd_pass = self._perform_forward_passes(batch)
 
+        if hasattr(self.model, 'module'):
```
Comment (stephenroller) [inline projects/anti_scaling/distillation.py:898]: Leave a comment please, even though this is common enough

Also maybe just roll onto one line:
`student_model = self.model.module if hasattr(self.model, 'module') else self.model`

File: tests/nightly/gpu/test_anti_scaling.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+            valid, _ = testing_utils.eval_model(Opt(opt), skip_test=True)
+            if not is_tinybert_style or precise_mode:
+                for loss_name, desired_loss in desired_losses.items():
+                    if np.isinf(desired_loss):
+                        self.assertTrue(np.isinf(valid[loss_name].value()))
+                    else:
+                        if abs(valid[loss_name].value() / desired_loss - 1) > 0.01:
+                            raise ValueError(
+                                f"""\
+Error in matching {loss_name} for {model_name}! \
```
Comment (stephenroller) [inline tests/nightly/gpu/test_anti_scaling.py:173]: Use multiline strings please, not \

Comment (stephenroller) [reply tests/nightly/gpu/test_anti_scaling.py:173]:
```python
ValueError(
    f'notice the lack of commas '
    f'or \\ for multiline strings'
)
```

File: tests/nightly/gpu/test_anti_scaling.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            model_name,
+            is_tinybert_style,
+            desired_losses,
+        ) in opts_and_desired_losses:
+            opt = {
+                **self.BASE_OPT,
+                **model_opt,
+                **distillation_opt,
+                'model': f'{self.DISTILLATION_MODEL_PREFIX}:{model_name}',
+                'num_examples': 1,
```
Comment (stephenroller) [inline tests/nightly/gpu/test_anti_scaling.py:157]: Given that you just want to run through a single example, let's remove the dependency on data.

Instead of using eval model, roll a manual observe act loop with a hardcoded text. Will be faster and more reliable.

Comment (stephenroller) [reply tests/nightly/gpu/test_anti_scaling.py:157]: You can get rid of the seeds too then.

For training, just set all the dropout parameters to 0.0 so it's deterministic.

File: tests/nightly/gpu/test_anti_scaling.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+                },
+            ),
+        ]
+        for (
+            model_opt,
+            distillation_opt,
+            model_name,
+            is_tinybert_style,
+            desired_losses,
+        ) in opts_and_desired_losses:
```
Comment (stephenroller) [inline tests/nightly/gpu/test_anti_scaling.py:151]: Looping in a test to enumerate is bad practice:
- it becomes hard to identify which sub-case broke
- i can't run only one sub-case that i'm trying to debug, and heaven forbid it's the last one that I need
- longer before I print something, increasing chance of a circleci timeout

File: tests/nightly/gpu/test_anti_scaling.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+                'pred_loss_coeff': 1,
+                'task_loss_coeff': 1,
+            }
+            valid, _ = testing_utils.eval_model(Opt(opt), skip_test=True)
+            if not is_tinybert_style or precise_mode:
+                for loss_name, desired_loss in desired_losses.items():
+                    if np.isinf(desired_loss):
+                        self.assertTrue(np.isinf(valid[loss_name].value()))
+                    else:
+                        if abs(valid[loss_name].value() / desired_loss - 1) > 0.01:
```
Comment (stephenroller) [inline tests/nightly/gpu/test_anti_scaling.py:170]: self.almostEquals

Comment (stephenroller) [reply tests/nightly/gpu/test_anti_scaling.py:170]: Given that you've specified your significant figures of losses much more than that, is that tolerance too liberal?

Maybe we should just move all of this to a pytest regression.

File: projects/anti_scaling/distillation.py [hunk 8]
```diff
@@ -0,0 +1,991 @@
...
+        num_tokens: torch.Tensor,
+        mapped_layers: List[int],
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        Compute the loss across all hidden layers for either the encoder or the decoder.
+
+        (The loss is averaged across all hidden layers and over the embedding dimension
+        so that it doesn't get too high for fp16 tensors.)
+        """
+        max_value = torch.finfo(student_hidden_states[0].dtype).max
```
[the remark this answers is not in the document]
Comment (stephenroller) [reply projects/anti_scaling/distillation.py:573]: I guess this is fine, it's just not very readable

File: parlai/agents/transformer/polyencoder.py [hunk 9]
```diff
@@ -411,7 +411,7 @@ def attend(self, attention_layer, queries, keys, values, mask):
         if isinstance(attention_layer, PolyBasicAttention):
             return attention_layer(queries, keys, mask_ys=mask, values=values)
         elif isinstance(attention_layer, MultiHeadAttention):
-            return attention_layer(queries, keys, values, mask)
+            return attention_layer(queries, keys, values, mask)[0]
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply parlai/agents/transformer/polyencoder.py:414]: Yeah, maybe I'm wrong, but I don't know how the code could have worked before - if you look at both places where `PolyEncoderModule.attend()` is called (lines 461 and 523), the result is expected to be a tensor, not a tuple

File: projects/anti_scaling/distillation.py [hunk 10]
```diff
@@ -0,0 +1,991 @@
...
+        )
+        self.record_local_metric(
+            'dec_hid_loss',
+            AverageMetric.many(
+                dec_hidden_loss_per_example, fwd_pass.tokens_per_example
+            ),
+        )
+        return enc_hidden_loss, dec_hidden_loss
+
+    @staticmethod
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:559]: Honestly, I don't have a strong preference one way or the other. I don't think there's any reason that they can't be regular non-static methods, so changing this to use self as the first arg as normal

File: projects/anti_scaling/distillation.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:573]: Good idea - switching to that utils.torch variable

File: projects/anti_scaling/distillation.py [hunk 11]
```diff
@@ -0,0 +1,991 @@
...
+            AverageMetric.many(pred_loss.sum(dim=-1), fwd_pass.tokens_per_example),
+        )
+        pred_loss = pred_loss.sum() / fwd_pass.num_tokens
+        return pred_loss
+
+
+class DistillTransformerAgentMixin(AbstractDistillTransformerAgentMixin):
+    @classmethod
+    def add_cmdline_args(cls, argparser):
+        super(DistillTransformerAgentMixin, cls).add_cmdline_args(argparser)
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:727]: yup, changed

File: projects/anti_scaling/distillation.py (hunk 2 above)
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py]: Ah yes - my thinking for this change was that, if you set `self.teacher_model` to the `'module'` attribute of the teacher in the __init__ method of the distillation agent, the effect of `DistributedDataParallel` may not get applied during training, but I don't know offhand if that's actually the case. This shift in the placement of this code also echos the fact that `self.model.module` in TGA is referenced in `TGA._generate()`, not `TGA.__init__()`

File: projects/anti_scaling/distillation.py (hunk 3 above)
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:898]: Definitely - added a comment. Just checked, and that code doesn't fit on one line

File: parlai/agents/transformer/modules.py [hunk 12]
```diff
@@ -1434,7 +1438,7 @@ def prepare_head(tensor):
 
         out = self.out_lin(attentioned)
 
-        return out, new_incr_state
+        return out, new_incr_state, dot_prod
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply parlai/agents/transformer/modules.py:1441]: We discussed this off-thread - I'll hold off on a flag, but I've gone back and double-checked all usages of `MultiHeadAttention` throughout the codebase

File: projects/anti_scaling/distillation.py [hunk 13]
```diff
@@ -0,0 +1,1022 @@
...
+        return hooks
+
+    @abstractmethod
+    def compute_loss(self, batch, return_output=False):
+        """
+        Return the loss.
+
+        This will likely call self._perform_forward_passes().
+        """
+        raise NotImplementedError('Abstract class: user must implement .compute_loss()')
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:274]: Changed, thanks!

File: projects/anti_scaling/scripts/remove_projection_matrices.py [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+    }
+    new_num_keys = len(states['model'])
+    print(f'{orig_num_keys-new_num_keys:d} model keys removed.')
+
+    print(f"Saving to {model_file}.")
+    atomic_save(states, model_file)
+
+
+if __name__ == '__main__':
+    random.seed(42)
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/scripts/remove_projection_matrices.py:46]: Whoops, removed

File: projects/anti_scaling/distillation.py [hunk 15]
```diff
@@ -0,0 +1,1022 @@
...
+            student_output=student_output,
+            student_scores=student_scores,
+            student_enc_output=student_enc_output,
+            student_embedding_outputs=student_embedding_outputs,
+            student_hidden_states=student_hidden_states,
+            student_attention_matrices=student_attention_matrices,
+        )
+
+
+class AbstractDistillTransformerAgentMixin(ABC):
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:117]: Hmm, the thing about that is then it makes solving "diamond problems" of multiple inheritance much harder - for example, `DistilBartAgent` inherits first from `DistilTransformerAgentMixin` and *then* from `BartAgent` (through `BartLikeAgent`), and if `DistilTransformerAgentMixin` *also* inherits from `TorchGeneratorAgent`, classes from `TorchGeneratorAgent` may override those of `BartAgent`. Although lemme know if you can think of an easy workaround :)

File: tests/nightly/gpu/test_anti_scaling.py [hunk 16]
```diff
@@ -0,0 +1,180 @@
...
+        student and teacher models is as expected.
+        """
+
+        precise_mode = False
+        # Turn this on to check the loss terms for TinyBERT-style distillation, which
+        # relies upon weights being initialized in a particular way. Won't work on
+        # CircleCI machines
+
+        torch.manual_seed(0)
+        np.random.seed(0)
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:79]: Yes, added

File: tests/nightly/gpu/test_anti_scaling.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        Check the sum of distillation losses.
+
+        Make sure that the sum of all distillation losses from one pass through the
+        student and teacher models is as expected.
+        """
+
+        precise_mode = False
+        # Turn this on to check the loss terms for TinyBERT-style distillation, which
+        # relies upon weights being initialized in a particular way. Won't work on
+        # CircleCI machines
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:76]: Hmm, not sure - even with the same seeds, I get different results on both boxes (but a consistent result on each individual box, of course). Maybe due to different underlying hardware?

File: tests/nightly/gpu/test_anti_scaling.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+from parlai.zoo.blender.blender_90M import download as download_blender
+
+
+class TestDistillation(unittest.TestCase):
+    """
+    Test agents for distilling transformer/generator models.
+    """
+
+    BLENDERBOT_MODEL_FILE = 'data/models/blender/blender_90M/model'
+    BART_MODEL_FILE = 'data/models/bart/bart_large/model'
```
[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:28]: Yes, good idea - will do

File: tests/nightly/gpu/test_anti_scaling.py (hunk 4 above)
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:173]: Whoops, yes, removing the 2 extra backslashes

File: tests/nightly/gpu/test_anti_scaling.py (hunk 5 above)
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:159]: Yes, good point - will replace with hardcoded text. However, I'll still need the seeds because the distillation agents still perform random initialization of weights, especially the TinyBERT-style agents where you can't copy over the teacher weights

File: tests/nightly/gpu/test_anti_scaling.py (hunk 6 above)
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:153]: Yes, will switch to one test per method

File: tests/nightly/gpu/test_anti_scaling.py (hunk 7 above)
Comment (EricMichaelSmith) [reply tests/nightly/gpu/test_anti_scaling.py:172]: hmm, yes, will move to pytest regression for this, good point

File: projects/anti_scaling/distillation.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (klshuster) [reply projects/anti_scaling/distillation.py:117]: that's a fair point, working with these inheritance chains can get confusing... maybe to solve at least the typing issues that are arising, you can add like `assert type(self) is TorchGeneratorAgent` to each function that makes a super call - this also would make the error more clear if for whatever reason the mixin is instantiated with the wrong agent type

[the remark this answers is not in the document]
Comment (EricMichaelSmith) [reply projects/anti_scaling/distillation.py:117]: Yes, good point - will add these to make sure those mypy warnings are quashed

**Comment by EricMichaelSmith** [comment]
@stephenroller @klshuster Thanks for many useful comments! I think that I've addressed them all now, but lemme know if you have other concerns with this

**Review by klshuster — APPROVED** [review]
lgtm

**Review by stephenroller — APPROVED** [review]
